x/txfees/keeper/hooks.go

```diff
@@ -0,0 +1,62 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	epochstypes "github.com/osmosis-labs/osmosis/v7/x/epochs/types"
+	txfeestypes "github.com/osmosis-labs/osmosis/v7/x/txfees/types"
+)
+
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) {}
+
+// at the end of each epoch, swap all non-OSMO fees into OSMO and transfer to fee module account
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) {
+	nonNativeFeeAddr := k.accountKeeper.GetModuleAddress(txfeestypes.NonNativeFeeCollectorName)
+	nonNativeBalances := k.bankKeeper.GetAllBalances(ctx, nonNativeFeeAddr)
+	baseDenom, _ := k.GetBaseDenom(ctx)
+
+	for _, coin := range nonNativeBalances {
```

Comment:
We should change the GetAllBalances and this for loop, to instead be iterating over every fee token denom.

This is because `GetAllBalances` can actually take unbounded time, and is not suitable for usage in the state machine

Reply:
Yes, this is a good point!